web/packages/teleterm/src/ui/services/connectionTracker/connectionTrackerService.test.tsx

```diff
@@ -0,0 +1,255 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import 'jest-canvas-mock';
+
+import { within } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { act, ComponentType, createRef } from 'react';
+
+import { render, screen } from 'design/utils/testing';
+import { App } from 'gen-proto-ts/teleport/lib/teleterm/v1/app_pb';
+
+import Logger, { NullService } from 'teleterm/logger';
+import { MockedUnaryCall } from 'teleterm/services/tshd/cloneableClient';
+import {
+  makeApp,
+  makeAppGateway,
+  makeRootCluster,
+} from 'teleterm/services/tshd/testHelpers';
+import { ResourcesContextProvider } from 'teleterm/ui/DocumentCluster/resourcesContext';
+import { MockAppContextProvider } from 'teleterm/ui/fixtures/MockAppContextProvider';
+import { MockAppContext } from 'teleterm/ui/fixtures/mocks';
+import { DocumentsService } from 'teleterm/ui/services/workspacesService';
+import { TabHost } from 'teleterm/ui/TabHost';
+import { IAppContext } from 'teleterm/ui/types';
+import { unique } from 'teleterm/ui/utils';
+
+import { TrackedGatewayConnection } from './types';
+
+beforeAll(() => {
+  Logger.init(new NullService());
+});
+
+test('updating target port creates new connection', async () => {
+  const user = userEvent.setup();
+  const { ctx, docsService, app, Component } = setupTests();
+
+  const doc1 = docsService.createGatewayDocument({
+    targetName: app.name,
+    targetUri: app.uri,
+    targetUser: undefined,
+    targetSubresourceName: '1337',
+    origin: 'resource_table',
+  });
+  // Add without opening. It's not necessary and it'll be easier to verify activating connections
+  // later if we don't open the doc at this point.
+  docsService.add(doc1);
+
+  render(<Component />);
+
+  // Wait for the gateway to be created.
+  expect(await screen.findByText('Close Connection')).toBeInTheDocument();
+  expect(ctx.connectionTracker.getConnections()).toHaveLength(1);
+  const conn1337 = ctx.connectionTracker.getConnections()[0];
+  expect(conn1337.title).toEqual(`${app.name}:1337`);
+
+  // Update target port.
+  let targetPortInput = screen.getByLabelText('Target Port');
+  await user.clear(targetPortInput);
+  await user.type(targetPortInput, '4242');
+  // We have to lose focus of that field, otherwise React is going to warn about updates not wrapped
+  // in act when the focus on the page changes after opening a new doc.
+  await user.tab();
```

Comment:
This threw me for a loop, omg. It might have to do with HTML validation or just focus changing, I haven't checked closely. I found info about `user.tab()` being useful here in a discussion under a library that we don't use. https://github.com/orgs/react-hook-form/discussions/4232#discussioncomment-8447514